### PR TITLE
Azimuth-tracker app

### PR DESCRIPTION
**Pull request by philipcmonk**
This is a new app called `azimuth-tracker` for the moment, which replaces Jael's functionality for tracking Ethereum nodes.  It is a port of @Fang- 's `:eth-watcher` app to the async framework, plus a new tracking algorithm.

The biggest change is that we now track each block, rather than trying to node-side log subscriptions.  I believe it's impossible to properly handle reorgs while depending on node-side state.  The basic algorithm is:

- walk-loop: for each block
  - fetch block from node
  - if parent of new block does not have same hash as the last block we processed locally, rewind-loop:
    - fetch previous block by number from node
    - if parent of new block has same hash as our local version of the previous block,
      - recurse to walk-loop
    - else, recurse to rewind-loop
  - else, release all events from the quarantine from at least 30 blocks ago
  - fetch new events for the current block (fetched by block hash, not number) and add them to the quarantine
  - recurse to walk-loop

Since this doesn't depend on node-side state, it happens to be Infura-safe, which is quite nice.

This is the correct algorithm, but it admits of some optimization.  For example, we request each block one at a time, but we could use a batch request to fetch 100 or 1000 at a time.  If we detect a reorg, we can walk backwards synchronously.

We may also want a "fast-forward" mode that uses getLogs with a range of blocks rather than a hash -- this is much faster, since it scales in the number of Azimuth events instead of Ethereum blocks, but it's not resilient to reorgs.  So maybe use that to catch up to within 1000 blocks, then switch to the true algorithm.

The app works, but it prints the updates to the screen rather than sending them to Jael/Kale, because I made a few changes to the types and haven't reconciled those with Kale.  I've tested this against real reorgs on the chain (1-block reorgs show up every 30 minutes or so).

Run with `:azimuth-tracker ['http://eth-mainnet.urbit.org:8545' 8.093.520]`.

Etc: I rather like this pattern of defining related types:

```
++  block
  =<  block
  |%
  +$  hash    @uxblockhash
  +$  number  @udblocknumber
  +$  id      [=hash =number]
  +$  block   [=id =parent=hash]
  --
```

Better than:

```
+$  block         [=block-id =parent=block-hash]
+$  block-hash    @uxblockhash
+$  block-number  @udblocknumber
+$  block-id      [=block-hash =block-number]
```

At call time, you can do:

```
|=  =block
=/  =our=id:block  `id:block`id.block
[parent-hash.block number.our-id]
```

Instead of:

```
|=  =block
=/  =our=block-id  `block-id`block-id.block
[parent-block-hash.blcok block-number.our-block-id]
```

Many fewer redundant "block" references, without polluting the global namespace at all.

### CODE REVIEW

**Comment by joemfb** [comment]
Also, there's a merge conflict with the pill. This is going to be an ongoing issue with managing pills in the repo, LFS commits obviously don't commute. Not sure how we should manage it, maybe resolving these just gets punted to @jtobin ;)

**Comment by ohAitch** [comment]
Ideally pills would continue to be built almost exclusively in CI, not sure
how easy this is to adapt to the LFS regime. Can Travis run on
merge-conflicting PRs? (I believe if one is opened by someone with org
commit privileges, Travis can transitively exercise the privileges for
linting and such.)

File: pkg/arvo/sys/zuse.hoon [hunk 1]
```diff
@@ -2313,21 +2313,112 @@
...
+        |=  diffs=(list diff)
+        ^-  (list diff)
+        %-  flop
+        %+  turn  diffs
+        |=  =diff
+        ^-  ^diff
+        ?-  -.diff
+          %rift  [%rift &2 |2]:diff
+          %keys  [%keys &2 |2]:diff
+          %spon  [%spon &2 |2]:diff
```
Comment (Fang-) [inline pkg/arvo/sys/zuse.hoon]: Is there a way to write this that's more common? This seems like tree addressing, but I haven't seen this notation before. (It's up in azimuth-tracker's `+parse-block` also, just noticed.)

File: pkg/arvo/app/azimuth-tracker.hoon [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+        ^-  form:m
+        ;<  eny=@uvJ  bind:m  get-entropy:stdio
+        ;<  now=@da   bind:m  get-time:stdio
+        %-  wait:stdio
+        %+  add  now
+        %+  min  limit
+        ?:  =(0 try)  ~s0
+        %+  add
+          (mul ~s1 (bex (dec try)))
+        (mul ~s0..0001 (~(rad og eny) 1.000))
```
[the remark this answers is not in the document]
Comment (philipcmonk) [reply pkg/arvo/app/azimuth-tracker.hoon]: This is copied verbatim from @joemfb 's DNS stuff.  In general, a little nondeterminism is healthy for backoff; not sure if there's something more specific here.

File: pkg/arvo/app/azimuth-tracker.hoon [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+      (pure:m udiffs)
+    ::
+    ++  jael-update
+      |=  =udiffs:point
+      =/  m  (async:stdio ,~)
+      |-  ^-  form:m
+      =*  loop  $
+      ?~  udiffs
+        (pure:m ~)
+      ~&  >  [%update block i.udiffs]
```
[the remark this answers is not in the document]
Comment (philipcmonk) [reply pkg/arvo/app/azimuth-tracker.hoon]: It's a syntax for "priority" printfs.  Also `>>` and `>>>`.  It used to just prepend the gars, but it looks like it doesn't anymore.

File: pkg/arvo/sys/zuse.hoon (hunk 1 above)
Comment (philipcmonk) [reply pkg/arvo/sys/zuse.hoon]: I'll use `+.diff`.

**Comment by philipcmonk** [comment]
The `~&`s are just because that's the only output right now -- it doesn't send anything anywhere else, so it's the only way to tell it's doing anything.  This isn't started automatically, so I think it's fine to leave them in for now.  I'll remove them when I integrate with Kale (in the next few days).

File: pkg/arvo/app/azimuth-tracker.hoon (hunk 2 above)
[the remark this answers is not in the document]
Comment (joemfb) [reply pkg/arvo/app/azimuth-tracker.hoon]: It's just adding a random number of milliseconds up to `~s1` -- you don't want things to accidentally correlate when backing off, and this function aims to be general-purpose.

File: pkg/arvo/app/azimuth-tracker.hoon (hunk 3 above)
[the remark this answers is not in the document]
Comment (joemfb) [reply pkg/arvo/app/azimuth-tracker.hoon]: It's still being printed (as in, there is code that prints it), but it looks like it's being swallowed in the terminal hijack/lojack stdout/stderr distinctions of the daemon. This stuff is a horrible mess, but I'll see if I can make this work again.

File: pkg/arvo/app/azimuth-tracker.hoon [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+      ::  ;<  ~  bind:m  (send-effect [%vent-update i.udiffs])
```
[the remark this answers is not in the document]
Comment (philipcmonk) [reply pkg/arvo/app/azimuth-tracker.hoon]: It should stay, but Kale doesn't actually handle this yet; am about to converge them.

**Comment by philipcmonk** [comment]
@jtobin this is ready to be merged

**Pull request merged**